charger/easee.go

```diff
@@ -306,6 +316,45 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
 	case easee.DYNAMIC_CHARGER_CURRENT:
 		c.dynamicChargerCurrent = value.(float64)
 	case easee.CHARGER_OP_MODE:
+		// New charging session pending, reset internal value of SESSION_ENERGY to 0, and its observation timestamp to "now".
+		// This should be done in a proper way by the api, but it's not.
+		// Remember value of LIFETIME_ENERGY as start value of the charging session
+		if c.opMode <= easee.ModeDisconnected && value.(int) >= easee.ModeAwaitingStart {
+			c.sessionEnergy = 0
+			c.obsTime[easee.SESSION_ENERGY] = time.Now()
+			c.sessionStartLifetimeEnergy = easee.NEED_SESSION_START_ENERGY
+		}
+
+		// OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
+		if c.opMode != easee.ModeCharging && value.(int) == easee.ModeCharging {
+			c.stopTicker = make(chan struct{})
```

Comment:
Ist sicher gestellt, dass der Channel hier immer nil ist bevor wir das machen? Mal probehalber mit einem panic versehen? Anderenfalls wäre das ein leak.

Reply:
@andig Ich hätte jetzt erwartet das mit Überschreiben von c.stopTicker ein möglicher vorhandener Channel einfach out of scope geht, und dann von der GC erfasst wird. Ist das ggf. bei Channels anders?

Reply:
Wenn er nicht closed ist läuft die Goroutine weiter und die verwendet den Channel. Wie stellen wir sicher, dass der Channel immer closed (oder geschrieben) ist bevor wir einen neuen schreiben?

Reply:
Danke! Jetzt sehe ich zumindest das Problem. Wenn wir OP_MODE updates out of order bekommen könnten wir hier landen bevor in 346 die Routine anhalten. Good catch!

Zur Lösung - könnten wir hier nicht einfach einen nil check einbauen, und nur bei != nil den Channel erstellen und die GoRoutine starten? Dazu dann noch wie unten angefordert in L346++ den Channel schließen und den Member auf nil setzen?

Reply:
Genau. Dann unten close und auf nil setzen, aber mit Mutex.

Reply:
Sollte nach meinem Verständnis nun hübsch sein.
Zusätzlich habe ich in Zeile 351 ein `c.stopTicker != nil` angehängt, denn denke mit der gleichen Logik, warum der channel hier noch != nil sein könnte
```
// OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
		if c.opMode != easee.ModeCharging && value.(int) == easee.ModeCharging {
			if c.stopTicker == nil {
				c.stopTicker = make(chan struct{})
			}
```
könnte er hier ja auch **noch nicht** != nil sein.
```
		// OpMode changed FROM charging to something else - stop ticker if channel exists
		if c.opMode == easee.ModeCharging && value.(int) != easee.ModeCharging && c.stopTicker != nil {
			c.stopTicker <- struct{}{}
			close(c.stopTicker)
			c.stopTicker = nil
		}

```

Wobei da die ProductUpdates aber wirklich sehr extrem durcheinander kommen müssten, insb. in Anbetracht des Filters auf das Alter. Aber schaden tut's ja nicht...